### PR TITLE
MAINT: import time: avoid repeated textwrap function dispatch instantiation

### PR DESCRIPTION
**Pull request by hmaarrfk**
Avoid the use of textwrap to deindent static code. This shaves about 6-8ms ms of import time (at the expense of slightly less readable code depending on who you ask -- but definitely if you ask @eric-wieser)

After: 99.1±2ms
Before: 106±2ms (master)

### CODE REVIEW

File: numpy/ma/core.py
Comment (hmaarrfk) [inline numpy/ma/core.py]: formatting definitely doesn't look right. give me a second.

**Comment by eric-wieser** [comment]
I'd be curious to know how much of this is the cost of importing `textwrap`, and how much is the cost of invoking `dedent`. I assume the stats you give above do not include the benefit of removing an import-time dependenct on `re`?

**Comment by hmaarrfk** [comment]
The stats I gave were obtained from running

```
asv continuous -b time_numpy  master import_time_textwrap -E conda:3.7
```

with this branch including only this one commit. As such, the code is still improting `re` elsewhere.

A runtime dependency on `re` is very difficult to remove. Let me try to find a concrete example.

**Comment by eric-wieser** [comment]
>  Let me try to find a concrete example.

No need, just wanted to confirm that this timing is not caused by removing it.

In the interest of breaking down the import cost here, can you leave behind the `import textwrap` but not actually use it, and see how that import time compares?

**Comment by hmaarrfk** [comment]
yeah, i just looked at the textwrap source, it seems pretty trivial, and shouldn't be a contributor, let me test.

**Comment by hmaarrfk** [comment]
It seems like it is the fact that we are using dedent that is causing the slowdown


```
$ asv continuous -b time_numpy  master import_time_textwrap -E conda:3.7
· `wheel_cache_size` has been renamed to `build_cache_size`. Update your `asv.conf.json` accordingly.
· No executable found for python 3.6
· Creating environments
· Discovering benchmarks
·· Uninstalling from conda-py3.7-six
·· Installing 690132df <import_time_textwrap> into conda-py3.7-six.
· Running 6 total benchmarks (2 commits * 1 environments * 3 benchmarks)
[  0.00%] · For numpy commit ea965e4c <master> (round 1/2):
[  0.00%] ·· Building for conda-py3.7-six.......
[  0.00%] ·· Benchmarking conda-py3.7-six
[  8.33%] ··· Running (bench_import.Import.time_numpy--)...
[ 25.00%] · For numpy commit 690132df <import_time_textwrap> (round 1/2):
[ 25.00%] ·· Building for conda-py3.7-six.
[ 25.00%] ·· Benchmarking conda-py3.7-six
[ 33.33%] ··· Running (bench_import.Import.time_numpy--)...
[ 50.00%] · For numpy commit 690132df <import_time_textwrap> (round 2/2):
[ 50.00%] ·· Benchmarking conda-py3.7-six
[ 58.33%] ··· bench_import.Import.time_numpy                                                       101±1ms
[ 66.67%] ··· bench_import.Import.time_numpy_inspect                                               104±1ms
[ 75.00%] ··· bench_linalg.Lstsq.time_numpy_linalg_lstsq_a__b_float64                          2.69±0.08ms
[ 75.00%] · For numpy commit ea965e4c <master> (round 2/2):
[ 75.00%] ·· Building for conda-py3.7-six.
[ 75.00%] ·· Benchmarking conda-py3.7-six
[ 83.33%] ··· bench_import.Import.time_numpy                                                       107±2ms
[ 91.67%] ··· bench_import.Import.time_numpy_inspect                                               109±2ms
[100.00%] ··· bench_linalg.Lstsq.time_numpy_linalg_lstsq_a__b_float64                          2.64±0.04ms

BENCHMARKS NOT SIGNIFICANTLY CHANGED.
(numpy) mark@mark-xps ~/g/n/benchmarks import_time_textwrap|+1⚑ 1
$ git log -n 1
commit 690132dfad9ddc775fd4607581050af048c9f84f (HEAD -> import_time_textwrap, origin/import_time_textwrap)
Author: Mark Harfouche <mark.harfouche@gmail.com>
Date:   Tue Jul 23 22:59:41 2019 -0400

    Import textwrap again just to test
```

File: numpy/ma/core.py
Comment (eric-wieser) [inline numpy/ma/core.py]: An alternative - `_legacy_print_templates` could also be put behind a function with `functools.lru_cache` - probably 99% of users never hit the code path that needs these.

File: numpy/core/overrides.py
Comment (eric-wieser) [inline numpy/core/overrides.py]: I suspect this is where all the cost is - we end up dedenting the same string for every single numpy function, which is obviously expensive. Dedenting it just once wouldn't be nearly as bad, although at that point chances are the string definition lives outside the function, at which point there's nothing to even dedent.

In the interest of preserving the visual indent, I might be inclined to move this string definition to a global `_wrapped_func_source` variable in the file, which still avoids paying for textwrap

**Comment by hmaarrfk** [comment]
@eric-wieser, you know, my original attempt at ripping out `re` used a ton of `lru_cache`. I felt like I would never code like that unless I was really trying to shave off milliseconds, and even then, it would end up being in vain since somebody else would end up importing `re` and the `lru_cache` would likely make things slower than the original case.

I think in this case, manual deindentenation is an appropriate compromise.

File: numpy/core/overrides.py
Comment (hmaarrfk) [reply numpy/core/overrides.py]: > In the interest of preserving the visual indent, I might be inclined to move this string definition to a global _wrapped_func_source variable in the file, which still avoids paying for textwrap

Is this an opinion, or a requirement to get your approval on this PR?

Comment (hmaarrfk) [reply numpy/core/overrides.py]: But I do suspect you are right in the source of the problem. I just don't personally see any issue with deindenting strings manually.

**Comment by hmaarrfk** [comment]
@eric-wieser. My milliseconds need this more than my intention to discuss coding style.

Benchmarks show similar improvements, again, hard to say when you are using the computer. Seeing as textwrap is built on top of `re`, it it likely only adds `1 ms` instead of the `6 ms` it added before.

Thanks for hunting down the true source of the issue. I won't be affraid to use textwrap in my other projects now.

**Review by eric-wieser — APPROVED** [review]
I could have been persuaded to take the original with some more thought, but this shortcuts that step. Might be worth a comment explaining it's location.

Can you confirm this still gives the savings you needed?

File: numpy/ma/core.py
Comment (eric-wieser) [reply numpy/ma/core.py]: For what it's worth, timing the difference between dedenting and not was ~0.1ms here 0.

Phrased more damningly, it was a factor of 100 slowdown, but on an already fast operation.

**Comment by hmaarrfk** [comment]
Yeah, it is still about a 4-6 ms in saving. Pretty good.

File: numpy/core/overrides.py
Comment (eric-wieser) [reply numpy/core/overrides.py]: > Is this an opinion

It's an opinion that probably gates my immediate approval until another maintainer weighs ins.

Comment (hmaarrfk) [reply numpy/core/overrides.py]: I'm over it, lets hunt more milliseconds, and work to deprecate the whole testing module. Much more interesting ;)

Comment (charris) [reply numpy/core/overrides.py]: I have a problem with dedenting strings manually, they are ugly, and ugly is unpleasant to maintain.

Comment (hmaarrfk) [reply numpy/core/overrides.py]: very fair.

**Comment by seberg** [comment]
I agree, simply moving out the string formatting seems fine for that import time boost, thanks @hmaarrfk. I will assume that even if we do things such as vendoring dedent this will still make sense. So putting it in.

File: numpy/core/overrides.py
Comment (seberg) [inline numpy/core/overrides.py]: Opst, empty line too much (don't worry about it though), just noticed that I did not nitpick ;).

Comment (hmaarrfk) [reply numpy/core/overrides.py]: Opps

**Pull request merged**
